Emby.Server.Implementations/QuickConnect/QuickConnectManager.cs

```diff
@@ -0,0 +1,296 @@
+        /// <inheritdoc/>
+        public string GenerateCode()
+        {
+            int min = (int)Math.Pow(10, CodeLength - 1);
+            int max = (int)Math.Pow(10, CodeLength);
+
+            uint scale = uint.MaxValue;
+            while (scale == uint.MaxValue)
+            {
+                byte[] raw = new byte[4];
+                _rng.GetBytes(raw);
+                scale = BitConverter.ToUInt32(raw, 0);
+            }
+
+            int code = (int)(min + ((max - min) * (scale / (double)uint.MaxValue)));
+            return code.ToString(CultureInfo.InvariantCulture);
+        }
+
+        /// <inheritdoc/>
+        public bool AuthorizeRequest(IRequest request, string code)
+        {
+            ExpireRequests();
+            AssertActive();
+
+            var auth = _authContext.GetAuthorizationInfo(request);
+
+            if (!_currentRequests.TryGetValue(code, out QuickConnectResult result))
+            {
+                throw new ResourceNotFoundException("Unable to find request");
+            }
+
+            if (result.Authenticated)
+            {
+                throw new InvalidOperationException("Request is already authorized");
+            }
+
+            result.Authentication = Guid.NewGuid().ToString("N", CultureInfo.InvariantCulture);
+
+            // Change the time on the request so it expires one minute into the future. It can't expire immediately as otherwise some clients wouldn't ever see that they have been authenticated.
+            var added = result.DateAdded ?? DateTime.Now.Subtract(new TimeSpan(0, Timeout, 0));
+            result.DateAdded = added.Subtract(new TimeSpan(0, Timeout - 1, 0));
+
+            _authenticationRepository.Create(new AuthenticationInfo
+            {
+                AppName = TokenNamePrefix + result.FriendlyName,
+                AccessToken = result.Authentication,
+                DateCreated = DateTime.UtcNow,
+                DeviceId = _appHost.SystemId,
+                DeviceName = _appHost.FriendlyName,
+                AppVersion = _appHost.ApplicationVersionString,
+                UserId = auth.UserId
+            });
+
+            _logger.LogInformation("Allowing device {0} to login as user {1} with quick connect code {2}", result.FriendlyName, auth.User.Username, result.Code);
+
+            return true;
+        }
+
+        /// <inheritdoc/>
+        public int DeleteAllDevices(Guid user)
+        {
+            var raw = _authenticationRepository.Get(new AuthenticationInfoQuery()
+            {
+                DeviceId = _appHost.SystemId,
+                UserId = user
+            });
+
+            var tokens = raw.Items.Where(x => x.AppName.StartsWith(TokenNamePrefix, StringComparison.CurrentCulture));
+
+            foreach (var token in tokens)
+            {
+                _authenticationRepository.Delete(token);
+                _logger.LogDebug("Deleted token {0}", token.AccessToken);
+            }
+
+            return tokens.Count();
```

Comment:
This iterates 2x over the `IEnumerable`, right?

Reply:
Yes, since I used the system id as the device id in the authentication info, I needed to do another pass over the returned tokens to avoid deleting any non quick connect authentication tokens.

Reply:
I suggest either using a counter, or use `.ToList()` to only iterate over the `IEnumerable` once